src/vdaf/suite.rs

```diff
@@ -0,0 +1,218 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module implements the cryptographic dependencies for our VDAF implementation. This
+//! includes a stream cipher (here we call it a [`KeyStream`]) and a pseudorandom function whose
+//! output size is the same as the key size (we call this a [`KeyDeriver`]).
+//!
+//! TODO(cjpatton) Add unit test with test vectors.
+
+use aes::cipher::generic_array::GenericArray;
+use aes::cipher::{FromBlockCipher, NewBlockCipher, StreamCipher as AesStreamCipher};
+use aes::{Aes128, Aes128Ctr};
+use getrandom::getrandom;
+use ring::hmac;
+use serde::{Deserialize, Serialize};
+
+const BLAKE3_DERIVE_PREFIX: &[u8] = b"blake3 key derive";
+const BLAKE3_STREAM_PREFIX: &[u8] = b"blake3 key stream";
+
+/// Errors propagated by methods in this module.
+#[derive(Debug, PartialEq, thiserror::Error)]
+pub enum SuiteError {
+    /// Failure when calling getrandom().
+    #[error("getrandom: {0}")]
+    GetRandom(#[from] getrandom::Error),
+}
+
+/// A suite uniquely determines a [`KeyStream`] and [`KeyDeriver`].
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Suite {
+    /// The [`KeyStream`] is implemented from AES128 in CTR mode and the [`KeyDeriver`] is
+    /// implemented from HMAC-SHA256.
+    Aes128CtrHmacSha256,
+
+    /// Both primitives are implemented using the BLAKE3 keyed hash function. The [`KeyStream']
+    /// uses the XOF mode and the [`KeyDeriver`] uses the standard, fixed-sized output mode.
+    Blake3,
+}
+
+/// A Key used to instantiate a [`KeyStream`] or [`KeyDeriver`].
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(untagged)]
```

Comment:
`untagged` makes Serde omit the variant name from the serialization. So if you have

```
struct {
    my_key: Key,
}
```

Then it would serialize to
```
{
    "my_key": {
         "Aes128CtrHmacSha256": [<32 bytes>]
    }
}
```
With flatten, it's just:
```
{
    "my_key": [<32 bytes>]
}
```
I think that's what we want: anybody serializing or deserializing a `Key` should know from context which suite is in use, and so encoding that on the wire is redundant.